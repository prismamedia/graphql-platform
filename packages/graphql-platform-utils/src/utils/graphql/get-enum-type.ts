import { GraphQLEnumType, GraphQLEnumValueConfigMap } from 'graphql';
import { POJO } from '../../types/pojo';

export function getGraphQLEnumType(name: string, enumerable: POJO, useKeyAsValue: boolean = false): GraphQLEnumType {
  return new GraphQLEnumType({
    name,
    values: Object.entries(enumerable).reduce(
      (values: GraphQLEnumValueConfigMap, [key, value]) =>
        !Number.isNaN(Number.parseInt(key))
          ? values
          : Object.assign(values, { [key]: { value: useKeyAsValue ? key : value } }),
      Object.create(null),
    ),
  });
}
