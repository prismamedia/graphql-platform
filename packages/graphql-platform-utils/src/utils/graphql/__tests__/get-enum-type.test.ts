import { getGraphQLEnumType } from '../get-enum-type';

enum ANumericEnum {
  Left,
  Right,
  Up,
  Down,
}

enum AStringEnum {
  Left = 'LEFT',
  Right = 'RIGHT',
  Up = 'UP',
  Down = 'DOWN',
}

describe('getGraphQLEnumType', () => {
  it('creates an enum type from numeric enum', () => {
    expect(getGraphQLEnumType('ANumericEnum', ANumericEnum).toConfig()).toMatchSnapshot();
  });

  it('creates an enum type from numeric enum using "key as value"', () => {
    expect(getGraphQLEnumType('ANumericEnumUsingKeyAsValue', ANumericEnum, true).toConfig()).toMatchSnapshot();
  });

  it('creates an enum type from string enum', () => {
    expect(getGraphQLEnumType('AStringEnum', AStringEnum).toConfig()).toMatchSnapshot();
  });

  it('creates an enum type from string enum using "key as value"', () => {
    expect(getGraphQLEnumType('AStringEnumUsingKeyAsValue', AStringEnum, true).toConfig()).toMatchSnapshot();
  });
});
