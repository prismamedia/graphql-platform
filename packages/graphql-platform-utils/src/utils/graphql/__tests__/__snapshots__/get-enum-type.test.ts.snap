// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGraphQLEnumType creates an enum type from numeric enum 1`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
  "name": "ANumericEnum",
  "values": Object {
    "Down": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": 3,
    },
    "Left": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": 0,
    },
    "Right": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": 1,
    },
    "Up": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": 2,
    },
  },
}
`;

exports[`getGraphQLEnumType creates an enum type from numeric enum using "key as value" 1`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
  "name": "ANumericEnumUsingKeyAsValue",
  "values": Object {
    "Down": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Down",
    },
    "Left": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Left",
    },
    "Right": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Right",
    },
    "Up": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Up",
    },
  },
}
`;

exports[`getGraphQLEnumType creates an enum type from string enum 1`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
  "name": "AStringEnum",
  "values": Object {
    "Down": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "DOWN",
    },
    "Left": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "LEFT",
    },
    "Right": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "RIGHT",
    },
    "Up": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "UP",
    },
  },
}
`;

exports[`getGraphQLEnumType creates an enum type from string enum using "key as value" 1`] = `
Object {
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
  "name": "AStringEnumUsingKeyAsValue",
  "values": Object {
    "Down": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Down",
    },
    "Left": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Left",
    },
    "Right": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Right",
    },
    "Up": Object {
      "astNode": undefined,
      "deprecationReason": undefined,
      "description": undefined,
      "extensions": undefined,
      "value": "Up",
    },
  },
}
`;
